dts: phy: Add clock-reference prop in stm32u5-otghs-phy

nandojve opened the pull request:
The OTG_HS PHY from `stm32u5a5xx` device require the correct reference clock frequency selction in `SYSCFG_OTGHSPHYCR`. The current default is hard coded to `16Mhz` (which matches the development board crystal). However, a custom board my require a different crystal and then the USB will not work. This add a required field in the `st,stm32u5-otghs-phy` binding to force user to select the correct clock reference. The current `nucleo_u5a5zj_q` baord was updated to reflect the mandatory field.

nandojve committed:
dts: phy: Add clock-reference prop in stm32u5-otghs-phy

The OTG_HS PHY from stm32u5a5xx device require the correct reference
clock frequency selction in SYSCFG_OTGHSPHYCR. The current default is
hard coded to 16Mhz (which matches the development board crystal).
However, a custom board my require a different crystal and then the
USB will not work. This add a required field in the
st,stm32u5-otghs-phy binding to force user to select the correct
clock reference. The current nucleo_u5a5zj_q baord was updated to
reflect the mandatory field.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

mathieuchopstm reviewed (dismissed):
\+ maybe an entry in migration guide should be added?

nandojve committed:
docs: migration-guide-4.3: Add st,stm32u5-otghs-phy note

Add migration note with details to guide user to correct select the
clock reference.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

dsseng commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -34,6 +34,14 @@ Device Drivers and Devicetree
+Phy
+=======
+
+* Nodes with compatible property :dtcompatible:`st,stm32u5-otghs-phy` now needs to select the
```
```suggestion
* Nodes with compatible property :dtcompatible:`st,stm32u5-otghs-phy` now need to select the
```

dsseng commented on dts/bindings/phy/st,stm32u5-otghs-phy.yaml:
```diff
@@ -36,3 +36,29 @@ properties:
+  clock-reference:
+    type: int
+    required: true
+    enum:
+      - 0x03
+      - 0x08
+      - 0x09
+      - 0x0a
+      - 0x0e
+      - 0x0b
```
Could this benefit from some form of constants to improve DT readability?

See this discussed for another driver: https://github.com/zephyrproject-rtos/zephyr/pull/87981#discussion_r2273814744

erwango reviewed (changes requested):
Thanks for tackling this

erwango reviewed (approved):
Approved.

Ideally we should also update migration guide, but error will be pretty explicit and binding is sufficiently documented to avoid a complicated journey to users.

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -67,6 +67,17 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #define USB_USBPHYC_CR_FSEL_24MHZ        USB_USBPHYC_CR_FSEL_1
 #endif
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_otghs) && defined(CONFIG_SOC_SERIES_STM32U5X)
+static const int syscfg_otg_hs_phy_clk[] = {
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_1,	/* 16Mhz   */
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_2,	/* 19.2Mhz */
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_3,	/* 20Mhz   */
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_4,	/* 24Mhz   */
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_5,	/* 26Mhz   */
+	SYSCFG_OTG_HS_PHY_CLK_SELECT_6,	/* 32Mhz   */
+};
+#endif
```
This is a bit wasteful. (c.f. other comment)

mathieuchopstm commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1063,7 +1074,9 @@ static int priv_clock_enable(void)
 	/* Set the OTG PHY reference clock selection (through SYSCFG) block */
 	LL_APB3_GRP1_EnableClock(LL_APB3_GRP1_PERIPH_SYSCFG);
-	HAL_SYSCFG_SetOTGPHYReferenceClockSelection(SYSCFG_OTG_HS_PHY_CLK_SELECT_1);
+	HAL_SYSCFG_SetOTGPHYReferenceClockSelection(
+		syscfg_otg_hs_phy_clk[DT_ENUM_IDX(DT_NODELABEL(otghs_phy), clock_reference)]
+	);
```
```suggestion
	HAL_SYSCFG_SetOTGPHYReferenceClockSelection(
		CONCAT(SYSCFG_OTG_HS_PHY_CLK_SELECT_, DT_ENUM_IDX(DT_NODELABEL(otghs_phy), clock_reference))
	);
```

mathieuchopstm reviewed (commented):
Not really blocking, but would be nice. Otherwise LGTM.

(Rationale: even though GCC ***might*** already optimize away the whole array stuff, I'd rather avoid relying on that)
